org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java

```diff
@@ -72,47 +94,174 @@ public CommandScope(Sig sig, boolean isExact, int scope) throws ErrorSyntax {
+    }
+
     /**
-     * Construct a new CommandScope object.
-     *
+     * The sig is still shallow copy.
+     * @return
+     */
+    public Object clone(){
+        return new CommandScope(this.pos, this.sig,this.isExact,this.startingScope,
+                this.endingScope,this.increment,new ArrayList<ExprVar>(this.pAtoms),
+                this.pAtomsLowerLastIndex, new ArrayList<List<Expr>>(this.pFields),
+                this.isPartial,this.hasLower,this.hasUpper,
+                this.isSparse);
+    }
+
+    /** Construct a new CommandScope object.
      * @param pos - the position where this scope is given
      * @param sig - the sig for this scope
      * @param isExact - true iff the scope is intended to be exact
      * @param startingScope - the starting scope
-     * @param endingScope - the ending scope (if this sig is not intended to be
-     *            growable, then startingScope should equal endingScope)
-     * @param increment - the scope increment (if this sig is not intended to be
-     *            growable, then this field is ignored)
+     * @param endingScope - the ending scope (if this sig is not intended to be growable, then startingScope should equal endingScope)
+     * @param increment - the scope increment (if this sig is not intended to be growable, then this field is ignored)
      * @throws ErrorSyntax if startingScope is less than zero
      * @throws ErrorSyntax if endingScope is less than startingScope
      * @throws ErrorSyntax if increment is less than one
      */
     public CommandScope(Pos pos, Sig sig, boolean isExact, int startingScope, int endingScope, int increment) throws ErrorSyntax {
-        if (pos == null)
-            pos = Pos.UNKNOWN;
-        if (sig == null)
-            throw new NullPointerException();
-        if (startingScope < 0)
-            throw new ErrorSyntax(pos, "Sig " + sig + " cannot have a negative starting scope (" + startingScope + ")");
-        if (endingScope < 0)
-            throw new ErrorSyntax(pos, "Sig " + sig + " cannot have a negative ending scope (" + endingScope + ")");
-        if (endingScope < startingScope)
-            throw new ErrorSyntax(pos, "Sig " + sig + " cannot have an ending scope (" + endingScope + ") smaller than its starting scope (" + startingScope + ")");
-        if (startingScope == endingScope)
-            increment = 1;
-        if (increment < 1)
-            throw new ErrorSyntax(pos, "Sig " + sig + "'s increment value cannot be " + increment + ".\nThe increment must be 1 or greater.");
+        this(pos, sig, isExact, startingScope, endingScope,increment,false);
+    }
+
+    /**
+     * It is similar the previous constructor, but sets the sparse boolean value to see whether integer values in inst-block should be included in the
+     * atoms or not. 
+     * @param pos
+     * @param sig
+     * @param isExact
+     * @param startingScope
+     * @param endingScope
+     * @param increment
```

Comment:
either add doc comments for these parameters or delete `@param` (just don't leave them empty)

Reply:
fixed